File: jetson/cfr_arduino_bridge/src/cloud_segmentation_node.py
#!/usr/bin/env python3
"""Run cloud_segmentation on the live ZED cloud and republish it by class.

Subscribes to the ZED's registered cloud and its pose, classifies every
point, and publishes the cloud again with each point's color set by its
class, so RViz and the browser viewer can show what the segmenter decided
rather than what the camera saw. The pose is only read for its pitch and
roll -- on the car the ZED's map frame is gravity-aligned, in the sim the
bridged ground-truth pose stands in for it -- which is what cloud_segmentation
needs to level the cloud.

Colors: ground gray-green, obstacle red, hoop magenta, car wash cyan,
overhead blue, unknown dark gray.
"""

import math

import numpy as np
import rclpy
from geometry_msgs.msg import PoseStamped
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import PointCloud2, PointField

import cloud_segmentation as cs

CLASS_COLORS = {
    cs.GROUND: (96, 128, 96),
    cs.OBSTACLE: (230, 50, 40),
    cs.HOOP: (240, 60, 240),
    cs.CARWASH: (40, 220, 230),
    cs.OVERHEAD: (70, 110, 255),
    cs.UNKNOWN: (60, 60, 60),
}


def packed_colors(labels: np.ndarray) -> np.ndarray:
    """uint32 0x00RRGGBB per point, the PCL/ROS packed-rgb convention."""
    table = np.zeros(256, dtype=np.uint32)
    for label, (r, g, b) in CLASS_COLORS.items():
        table[label] = (r << 16) | (g << 8) | b
    return table[labels]


def xyz_from_cloud(msg: PointCloud2) -> np.ndarray:
    offsets = {f.name: f.offset for f in msg.fields}
    count = msg.width * msg.height
    raw = np.frombuffer(msg.data, dtype=np.uint8, count=count * msg.point_step)
    raw = raw.reshape(count, msg.point_step)
    endian = ">f4" if msg.is_bigendian else "<f4"
    return np.stack(
        [
            raw[:, offsets[a] : offsets[a] + 4].copy().view(endian).ravel()
            for a in "xyz"
        ],
        axis=1,
    )


def pitch_roll(q) -> tuple[float, float]:
    """Nose-down pitch and left-up roll from a REP-103 orientation."""
    pitch = math.asin(max(-1.0, min(1.0, 2.0 * (q.w * q.y - q.z * q.x))))
    roll = math.atan2(
        2.0 * (q.w * q.x + q.y * q.z), 1.0 - 2.0 * (q.x * q.x + q.y * q.y)
    )
    return pitch, roll


class CloudSegmentation(Node):
    def __init__(self):
        super().__init__("cloud_segmentation")
        # Segmentation runs on the full cloud -- thin structure at range is
        # what the segmenter needs every point for -- and only what is
        # published is thinned, to every stride-th row and column, because
        # the browser draws it and a quarter of 230k points is plenty to see.
        self.stride = int(self.declare_parameter("stride", 2).value)
        self.tilt = (0.0, 0.0)
        self.publisher = self.create_publisher(PointCloud2, "~/cloud", 1)
        self.create_subscription(PoseStamped, "pose", self.on_pose, 10)
        self.create_subscription(
            PointCloud2, "cloud", self.on_cloud, qos_profile_sensor_data
        )

    def on_pose(self, msg: PoseStamped) -> None:
        self.tilt = pitch_roll(msg.pose.orientation)

    def on_cloud(self, msg: PointCloud2) -> None:
        points = xyz_from_cloud(msg)
        labels = cs.segment(points, *self.tilt).labels
        if msg.height > 1 and self.stride > 1:
            grid = (msg.height, msg.width)
            points = points.reshape(*grid, 3)[:: self.stride, :: self.stride].reshape(
                -1, 3
            )
            labels = labels.reshape(grid)[:: self.stride, :: self.stride].ravel()
        keep = labels != cs.UNKNOWN
        points = points[keep].astype(np.float32)
        colors = packed_colors(labels[keep])

        out = PointCloud2()
        out.header = msg.header
        out.height = 1
        out.width = len(points)
        out.fields = [
            PointField(name="x", offset=0, datatype=PointField.FLOAT32, count=1),
            PointField(name="y", offset=4, datatype=PointField.FLOAT32, count=1),
            PointField(name="z", offset=8, datatype=PointField.FLOAT32, count=1),
            PointField(name="rgb", offset=12, datatype=PointField.FLOAT32, count=1),
        ]
        out.is_bigendian = False
        out.point_step = 16
        out.row_step = 16 * len(points)
        out.is_dense = True
        packed = np.empty((len(points), 4), dtype="<u4")
        packed[:, :3] = points.view("<u4")
        packed[:, 3] = colors
        out.data = packed.tobytes()
        self.publisher.publish(out)


def main():
    rclpy.init()
    node = CloudSegmentation()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()

File: jetson/cfr_arduino_bridge/src/cloud_segmentation_node.cpp
// Run cloud_segmentation on the live ZED cloud and republish it by class.
//
// Subscribes to the ZED's registered cloud and its pose, classifies every
// point, and publishes the cloud again with each point's color set by its
// class, so RViz and the browser viewer can show what the segmenter decided
// rather than what the camera saw.  The pose is only read for its pitch and
// roll -- on the car the ZED's map frame is gravity-aligned, in the sim the
// bridged ground-truth pose stands in for it -- which is what the segmenter
// needs to level the cloud.
//
// Colors: ground gray-green, obstacle red, hoop magenta, car wash cyan,
// overhead blue; unknown points are left out.
//
// | Interface | Type | Direction |
// | --------- | ---- | --------- |
// | `cloud` | `sensor_msgs/PointCloud2` | subscribed |
// | `pose` | `geometry_msgs/PoseStamped` | subscribed |
// | `~/cloud` | `sensor_msgs/PointCloud2` | published, x/y/z/rgb |
//
// Parameters: `stride` thins what is published (not what is segmented) to
// every stride-th row and column, because the browser draws it and a quarter
// of 230k points is plenty to see.  Every cloud_segmentation::Params field is
// a parameter too, under its own name, read at startup.  `cloud_reliable`
// (default true) subscribes reliably; see CloudQoS.

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <cstring>
#include <memory>
#include <vector>

#include "cfr_arduino_bridge/cloud_msg.hpp"
#include "cfr_arduino_bridge/cloud_segmentation.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/point_cloud2.hpp"
#include "sensor_msgs/msg/point_field.hpp"

namespace cfr_arduino_bridge {

  namespace seg = segmentation;

  // uint32 0x00RRGGBB per class, the PCL/ROS packed-rgb convention.
  inline uint32_t ClassColor(uint8_t label) {
    const auto rgb = [](uint32_t r, uint32_t g, uint32_t b) { return (r << 16) | (g << 8) | b; };
    switch (label) {
      case seg::kGround:
        return rgb(96, 128, 96);
      case seg::kObstacle:
        return rgb(230, 50, 40);
      case seg::kHoop:
        return rgb(240, 60, 240);
      case seg::kCarWash:
        return rgb(40, 220, 230);
      case seg::kOverhead:
        return rgb(70, 110, 255);
      default:
        return rgb(60, 60, 60);
    }
  }

  class CloudSegmentationNode : public rclcpp::Node {
   public:
    CloudSegmentationNode() : rclcpp::Node("cloud_segmentation") {
      stride_ = std::max<int64_t>(1, declare_parameter<int64_t>("stride", 2));
      const bool reliable = declare_parameter<bool>("cloud_reliable", true);
#define CFR_DECLARE(field) \
  params_.field = static_cast<decltype(params_.field)>(declare_parameter(#field, params_.field));
      CFR_SEGMENTATION_PARAMS(CFR_DECLARE)
#undef CFR_DECLARE

      publisher_ = create_publisher<sensor_msgs::msg::PointCloud2>("~/cloud", 1);
      pose_sub_ = create_subscription<geometry_msgs::msg::PoseStamped>(
          "pose", 10, [this](const geometry_msgs::msg::PoseStamped& msg) { OnPose(msg); });
      // Depth 1: a frame that waited behind a slow one is stale by the time
      // it is read, so only the newest is kept.
      cloud_sub_ = create_subscription<sensor_msgs::msg::PointCloud2>(
          "cloud", CloudQoS(reliable), [this](const sensor_msgs::msg::PointCloud2& msg) { OnCloud(msg); });
    }

   private:
    // Nose-down pitch and left-up roll from a REP-103 orientation.
    void OnPose(const geometry_msgs::msg::PoseStamped& msg) {
      const auto& q = msg.pose.orientation;
      pitch_ = std::asin(std::clamp(2.0 * (q.w * q.y - q.z * q.x), -1.0, 1.0));
      roll_ = std::atan2(2.0 * (q.w * q.x + q.y * q.z), 1.0 - 2.0 * (q.x * q.x + q.y * q.y));
    }

    void OnCloud(const sensor_msgs::msg::PointCloud2& msg) {
      CloudLayout layout;
      try {
        layout = LayoutOf(msg);
      } catch (const std::invalid_argument& error) {
        RCLCPP_ERROR_THROTTLE(get_logger(), *get_clock(), 5000, "cannot segment this cloud: %s", error.what());
        return;
      }
      const size_t n = layout.Count();
      xyz_.resize(3 * n);
      ForEachPoint(msg.data.data(), layout, [this](size_t i, float x, float y, float z) {
        xyz_[3 * i] = x;
        xyz_[3 * i + 1] = y;
        xyz_[3 * i + 2] = z;
      });
      // Both the ZED and Gazebo publish registered color, but their extra
      // point fields are not laid out identically. Accept packed rgb or rgba
      // in either FLOAT32 or UINT32 form, and keep geometry-only behavior if
      // the cloud does not provide color.
      const sensor_msgs::msg::PointField* color_field = nullptr;
      for (const auto& field : msg.fields) {
        if ((field.name == "rgb" || field.name == "rgba") &&
            (field.datatype == sensor_msgs::msg::PointField::FLOAT32 ||
             field.datatype == sensor_msgs::msg::PointField::UINT32) &&
            field.offset + 4 <= layout.point_step) {
          color_field = &field;
          break;
        }
      }
      if (color_field) {
        rgb_.resize(n);
        for (size_t i = 0; i < n; ++i) {
          const uint8_t* bytes = msg.data.data() + layout.Offset(i) + color_field->offset;
          rgb_[i] = msg.is_bigendian ? (uint32_t(bytes[0]) << 24) | (uint32_t(bytes[1]) << 16) |
                                           (uint32_t(bytes[2]) << 8) | uint32_t(bytes[3]) :
                                       (uint32_t(bytes[3]) << 24) | (uint32_t(bytes[2]) << 16) |
                                           (uint32_t(bytes[1]) << 8) | uint32_t(bytes[0]);
        }
      }
      seg::Segment(xyz_.data(), n, pitch_, roll_, params_, &result_, color_field ? rgb_.data() : nullptr);

      // Everything is segmented; only every stride-th row and column of an
      // organized cloud is published.
      const size_t stride = layout.height > 1 ? static_cast<size_t>(stride_) : 1;
      auto out = std::make_unique<sensor_msgs::msg::PointCloud2>();
      out->header = msg.header;
      out->height = 1;
      out->is_bigendian = false;
      out->point_step = 16;
      out->is_dense = true;
      const char* names[] = {"x", "y", "z", "rgb"};
      for (uint32_t k = 0; k < 4; ++k) {
        sensor_msgs::msg::PointField field;
        field.name = names[k];
        field.offset = 4 * k;
        field.datatype = sensor_msgs::msg::PointField::FLOAT32;
        field.count = 1;
        out->fields.push_back(field);
      }
      out->data.resize(16 * ((layout.height + stride - 1) / stride) * ((layout.width + stride - 1) / stride));
      uint8_t* write = out->data.data();
      for (size_t row = 0; row < layout.height; row += stride) {
        for (size_t col = 0; col < layout.width; col += stride) {
          const size_t i = row * layout.width + col;
          const uint8_t label = result_.labels[i];
          if (label == seg::kUnknown) {
            continue;
          }
          const float point[3] = {static_cast<float>(xyz_[3 * i]),
                                  static_cast<float>(xyz_[3 * i + 1]),
                                  static_cast<float>(xyz_[3 * i + 2])};
          const uint32_t color = ClassColor(label);
          std::memcpy(write, point, 12);
          std::memcpy(write + 12, &color, 4);
          write += 16;
        }
      }
      const size_t count = static_cast<size_t>(write - out->data.data()) / 16;
      out->data.resize(16 * count);
      out->width = static_cast<uint32_t>(count);
      out->row_step = static_cast<uint32_t>(16 * count);
      publisher_->publish(std::move(out));
    }

    int64_t stride_ = 2;
    seg::Params params_;
    double pitch_ = 0.0;
    double roll_ = 0.0;
    std::vector<double> xyz_;
    std::vector<uint32_t> rgb_;
    seg::Segmentation result_;
    rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr publisher_;
    rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_sub_;
    rclcpp::Subscription<sensor_msgs::msg::PointCloud2>::SharedPtr cloud_sub_;
  };

}  // namespace cfr_arduino_bridge

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<cfr_arduino_bridge::CloudSegmentationNode>());
  rclcpp::shutdown();
  return 0;
}
